compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs

```diff
@@ -0,0 +1,382 @@
+use rustc_index::bit_set::{BitSet, ChunkedBitSet};
+use rustc_macros::LintDiagnostic;
+use rustc_middle::mir::{
+    BasicBlock, Body, ClearCrossCrate, Local, Location, Place, ProjectionElem, StatementKind,
+    TerminatorKind, dump_mir,
+};
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_mir_dataflow::impls::MaybeInitializedPlaces;
+use rustc_mir_dataflow::move_paths::{MoveData, MovePathIndex};
+use rustc_mir_dataflow::{Analysis, MaybeReachable};
+use rustc_session::lint;
+use rustc_span::Span;
+use rustc_type_ir::data_structures::IndexMap;
+use smallvec::{SmallVec, smallvec};
+use tracing::{debug, instrument};
+
+fn place_has_common_prefix<'tcx>(left: &Place<'tcx>, right: &Place<'tcx>) -> bool {
+    left.local == right.local
+        && left.projection.iter().zip(right.projection).all(|(left, right)| left == right)
+}
+
+fn drops_reachable_from_location<'tcx>(
+    body: &Body<'tcx>,
+    block: BasicBlock,
+    place: &Place<'tcx>,
+) -> BitSet<BasicBlock> {
+    let mut reachable = BitSet::new_empty(body.basic_blocks.len());
+    let mut visited = reachable.clone();
+    let mut new_blocks = reachable.clone();
+    new_blocks.insert(block);
+    while !new_blocks.is_empty() {
+        let mut next_front = BitSet::new_empty(new_blocks.domain_size());
+        for bb in new_blocks.iter() {
+            if !visited.insert(bb) {
+                continue;
+            }
+            for succ in body.basic_blocks[bb].terminator.iter().flat_map(|term| term.successors()) {
+                let target = &body.basic_blocks[succ];
+                if target.is_cleanup {
+                    continue;
+                }
+                if !next_front.contains(succ)
+                    && let Some(terminator) = &target.terminator
+                    && let TerminatorKind::Drop {
+                        place: dropped_place,
+                        target: _,
+                        unwind: _,
+                        replace: _,
+                    } = &terminator.kind
+                    && place_has_common_prefix(dropped_place, place)
+                {
+                    reachable.insert(succ);
+                    // Now we have discovered a simple control flow path from a future drop point
+                    // to the current drop point.
+                    // We will not continue here.
+                } else {
+                    next_front.insert(succ);
+                }
+            }
+        }
+        new_blocks = next_front;
+    }
+    reachable
+}
+
+#[instrument(level = "debug", skip(tcx, param_env))]
+fn extract_component_raw<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    ty: Ty<'tcx>,
+) -> SmallVec<[Ty<'tcx>; 4]> {
+    // Droppiness does not depend on regions, so let us erase them.
+    let ty = tcx.try_normalize_erasing_regions(param_env, ty).unwrap_or(ty);
+
+    let Ok(tys) = tcx.list_significant_drop_tys(param_env.and(ty)) else {
+        return smallvec![ty];
+    };
```

Comment:
```suggestion
    let tys = tcx.list_significant_drop_tys(param_env.and(ty));
```

This can be simplified if you take my change to the query definition.

Reply:
Applied